### PR TITLE
Fix newline inserted in empty block at end of formatting range

### PR DESCRIPTION
**Pull request by andrewbranch**
Fixes bug tracked in Azure DevOps found in VS testing. Thanks @zkat for helping figure out a test repro.

### CODE REVIEW

File: src/services/formatting/formatting.ts [hunk 1]
```diff
@@ -439,20 +439,21 @@ namespace ts.formatting {
...
-                    sourceFile.getLineAndCharacterOfPosition(token.pos).line,
-                    enclosingNode,
+                    tokenInfo,
+                    sourceFile.getLineAndCharacterOfPosition(tokenInfo.pos).line,
+                    parent,
                     previousRange,
                     previousRangeStartLine!,
                     previousParent!,
-                    enclosingNode,
+                    parent,
```
Comment (andrewbranch) [inline src/services/formatting/formatting.ts:456]: The TLDR is this final parameter is supposed to be the “context node” of the token, as if anyone knows what that means, and `enclosingNode` may be arbitrarily higher up the tree. Turns out that “context node” means “direct parent” in almost all cases. What was happening in this case was that `contextNode` should have been the inner `if` Block (`{}`), but it was actually the whole outer `IfStatement` (`if (foo) { ... }`). The rule `NewLineBeforeCloseBraceInBlockContext` is supposed to take effect only if the block ending in the close brace is not on a single line. The inner one is, so the rule should not have applied, but it was using the outer one to decide whether to apply the rule.

File: src/services/utilities.ts [hunk 2]
```diff
@@ -1256,8 +1256,10 @@ namespace ts {
      * Finds the rightmost token satisfying `token.end <= position`,
      * excluding `JsxText` tokens containing only whitespace.
      */
-    export function findPrecedingToken(position: number, sourceFile: SourceFile, startNode?: Node, excludeJsdoc?: boolean): Node | undefined {
-        const result = find(startNode || sourceFile);
+    export function findPrecedingToken(position: number, sourceFile: SourceFileLike, startNode: Node, excludeJsdoc?: boolean): Node | undefined;
+    export function findPrecedingToken(position: number, sourceFile: SourceFile, startNode?: Node, excludeJsdoc?: boolean): Node | undefined;
```
Comment (jakebailey) [inline src/services/utilities.ts:1260]: Out of curiosity, why couldn't this just change to `SourceFileLike`? Seems like it'd be a backwards compatible change to accept a wider set of types.

Comment (jakebailey) [reply src/services/utilities.ts:1260]: Oh, I didn't notice that `startNode` becomes required.

**Review by jakebailey — APPROVED** [review]
LGTM

File: src/services/utilities.ts (hunk 2 above)
Comment (andrewbranch) [reply src/services/utilities.ts:1260]: `SourceFileLike` has no parse tree structure; it’s just a text container. So if you provide a `SourceFileLike` you _must_ provide an actual Node to start in.

Comment (jakebailey) [reply src/services/utilities.ts:1260]: Yeah, I just didn't notice the one character `?` difference, oops.

**Comment by DanielRosenwasser** [comment]
@typescript-bot cherry-pick this to release-4.6

**Pull request merged**